android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/Keyboard.java

```diff
@@ -1,90 +1,119 @@
-  }
+      // language ID and language name from languageJSON
+      this.languageID = languageJSON.getString(KMManager.KMKey_ID).toLowerCase();
+      this.languageName = languageJSON.getString(KMManager.KMKey_Name);
 
-  public String getLanguageName() {
-    return this.map.get(KMManager.KMKey_LanguageName);
-  }
+      this.isCustomKeyboard = keyboardJSON.has(KMManager.KMKey_CustomKeyboard) &&
+        keyboardJSON.get(KMManager.KMKey_CustomKeyboard).equals("Y");
 
-  public String getResourceName() {
-    return this.map.get(KMManager.KMKey_KeyboardName);
-  }
+      this.isNewKeyboard = keyboardJSON.has(KeyboardPickerActivity.KMKEY_INTERNAL_NEW_KEYBOARD) &&
+        keyboardJSON.get(KeyboardPickerActivity.KMKEY_INTERNAL_NEW_KEYBOARD).equals(KeyboardPickerActivity.KMKEY_INTERNAL_NEW_KEYBOARD);
+
+      this.font = keyboardJSON.optString(KMManager.KMKey_Font, "");
+
+      this.oskFont = keyboardJSON.optString(KMManager.KMKey_OskFont, "");
 
-  public String getCustomHelpLink() {
-    if (this.map.containsKey(KMManager.KMKey_CustomHelpLink)) {
-      return this.map.get(KMManager.KMKey_CustomHelpLink);
+      this.version = keyboardJSON.optString(KMManager.KMKey_KeyboardVersion, "1.0");
+
+      this.helpLink = keyboardJSON.optString(KMManager.KMKey_CustomHelpLink,
+        String.format(HELP_URL_FORMATSTR, this.keyboardID, this.version));
+    } catch (JSONException e) {
+      Log.e(TAG, "Keyboard exception parsing JSON: " + e);
     }
-    return null;
   }
 
-  public String getVersion() {
-    return this.map.get(KMManager.KMKey_KeyboardVersion);
+  public Keyboard(String packageID, String keyboardID, String keyboardName, String languageID, String languageName,
+                  boolean isCustomKeyboard, boolean isNewKeyboard,
+                  String font, String oskFont, String version, String helpLink) {
+
+    this.packageID = (packageID != null) ? packageID : KMManager.KMDefault_UndefinedPackageID;
+    this.keyboardID = keyboardID;
+    this.keyboardName = keyboardName;
+    this.languageID = languageID.toLowerCase();
+    this.languageName = languageName;
+    this.isCustomKeyboard = isCustomKeyboard;
+    this.isNewKeyboard = isNewKeyboard;
+    this.font = (font != null) ? font : "";
+    this.oskFont = (oskFont != null) ? oskFont : "";
+    this.version = (version != null) ? version : "1.0";
+    this.helpLink = (FileUtils.isWelcomeFile(helpLink)) ? helpLink :
+      String.format(HELP_URL_FORMATSTR, this.keyboardID, this.version);
   }
 
-  public String getPackage() {
-    return this.map.get(KMManager.KMKey_PackageID);
-  }
+  public boolean isCustomKeyboard() { return isCustomKeyboard; }
+
+  public boolean isNewKeyboard() { return isNewKeyboard; }
+
+  public String getResourceId() { return keyboardID; }
+
+  public String getFont() { return font; }
+
+  public String getOSKFont() { return oskFont; }
+
+  @Override
+  public String getLanguageCode() { return languageID; }
```

Comment:
Do you think we can offer also `getLanguageID()` and mark `getLanguageCode()` as deprecated? It's just a bit more consistent.

Reply:
I like that suggestion.

@jahorton , do you happen to remember why you chose `getLanguageCode()` and `getPackage()` in the base class android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/LanguageResource.java ?
See https://github.com/keymanapp/keyman/commit/95b5980915b1f54d680514be4a3ea26ef1613e06#diff-292a9e83463f518fed6fee36afdad0e2

Reply:
When I do the refactor for data/LexicalModel.java, I also intend to convert LanguageResource.java from interface to abstract

Reply:
I'm going to guess it was b/c of KeymanWeb - it's always used "language code" terminology, not "language ID".